### PR TITLE
Test LoanOffer and upgrade Charta

### PR DESCRIPTION
**Pull request by chrismin**

### CODE REVIEW

File: __test__/integration/loan_offer/runners/accept.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            collateralToken = await DummyTokenContract.at(
+                (await contractsApi.loadTokenBySymbolAsync(params.collateralToken)).address,
+                web3,
+                TX_DEFAULTS,
+            );
+        });
+
+        beforeEach(async () => {
+            // set balances
+            await principalToken.setBalance.sendTransactionAsync(
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/accept.ts]: similar to what we do in `loan`, let's create a function that sets these balances and allowances for us.

```export async function setBalancesAndAllowances(
    dharma: Dharma,
    params: LoanRequestParams,
    debtor: string,
    creditor: string,
): Promise<void> {
    // Configure balances and allowances for debtor.
    await setBalanceForSymbol(dharma, params.collateralAmount, params.collateralToken, debtor);
    await setUnlimitedAllowanceForSymbol(dharma, params.collateralToken, debtor);

    // Configure balances and allowances for creditor.
    await setBalanceForSymbol(dharma, params.principalAmount, params.principalToken, creditor);
    await setUnlimitedAllowanceForSymbol(dharma, params.principalToken, creditor);
}
```

File: __test__/integration/loan_offer/runners/accept.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    signAndAccept: (loanOffer: LoanOffer, address: string) => void,
+) {
+    describe("passing valid params", () => {
+        let loanOffer: LoanOffer;
+
+        let web3: Web3;
+
+        let contractsApi: ContractsAPI;
+
+        let principalToken: DummyTokenContract;
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/accept.ts]: we can remove a lot of this setup code by using a shared util function. see below.

File: __test__/integration/loan_offer/runners/accept.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+const CREDITOR = ACCOUNTS[0];
+const DEBTOR = ACCOUNTS[1];
+
+const TX_DEFAULTS = { from: CREDITOR.address, gas: 4712388 };
+
+export async function testAccept(
+    dharma: Dharma,
+    params: any,
+    signAndAccept: (loanOffer: LoanOffer, address: string) => void,
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/accept.ts]: address here is vague. who is this? debtor? creditor?

File: __test__/integration/loan_offer/runners/accept_as_debtor.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { Dharma } from "../../../../src";
+
+import { LoanOffer } from "../../../../src/types";
+
+import { testAccept } from "./accept";
+
+export async function testAcceptAsDebtor(dharma: Dharma, params: any) {
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/accept_as_debtor.ts]: any isn't a great type.

File: __test__/integration/loan_offer/runners/accept_as_debtor.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    async function signAndAccept(loanOffer: LoanOffer, address: string) {
+        await loanOffer.signAsDebtor(address);
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/accept_as_debtor.ts]: shouldn't need to sign and then accept

File: __test__/integration/loan_offer/runners/sign_as_creditor.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { Dharma } from "../../../../src";
+
+import { LoanOffer } from "../../../../src/types";
+
+export async function testSignAsCreditor(dharma: Dharma, params: any) {
```
Comment (saturnial) [inline __test__/integration/loan_offer/runners/sign_as_creditor.ts]: params should conform to a type

File: src/loan/loan_request.ts [hunk 7]
```diff
@@ -66,19 +66,6 @@ export class LoanRequest extends DebtOrder {
...
-    /**
-     * Eventually returns true if the current loan request has been filled by a creditor.
-     *
-     * @example
-     * await loanRequest.isFilled();
-     * => true
-     *
-     * @returns {Promise<boolean>}
-     */
-    public async isFilled(): Promise<boolean> {
```
Comment (saturnial) [inline src/loan/loan_request.ts]: `isFilled` should stay here.

File: src/loan/debt_order.ts [hunk 8]
```diff
@@ -295,6 +308,7 @@ export class DebtOrder {
      * @return {boolean}
      */
     public isSignedByDebtor(): boolean {
+        // TODO: check validity of signature
```
Comment (saturnial) [inline src/loan/debt_order.ts]: this is a great to do -- let's do it!

File: src/loan/debt_order.ts [hunk 9]
```diff
@@ -285,6 +285,19 @@ export class DebtOrder {
         return expirationTimestamp.lt(approximateNextBlockTime);
     }
 
+    /**
```
Comment (saturnial) [inline src/loan/debt_order.ts]: this shouldn't live on the debt order

File: src/types/loan_offer/loan_offer.ts [hunk 10]
```diff
@@ -85,20 +88,23 @@ export class LoanOffer extends DebtOrder {
      *
      * @return {Promise<string>}
      */
-    public async acceptAsProxy(proxyAddress?: string): Promise<string> {
+    public async acceptAsProxy(
+        proxyAddress?: string,
+        transactionOptions?: TxData,
```
[the remark this answers is not in the document]
Comment (chrismin) [reply src/types/loan_offer/loan_offer.ts]: Done.

File: src/loan/loan_request.ts (hunk 7 above)
Comment (chrismin) [reply src/loan/loan_request.ts]: Done.

File: src/loan/debt_order.ts (hunk 9 above)
Comment (chrismin) [reply src/loan/debt_order.ts]: Done.

File: src/loan/debt_order.ts (hunk 8 above)
Comment (chrismin) [reply src/loan/debt_order.ts]: done

File: __test__/integration/loan_offer/runners/sign_as_creditor.ts (hunk 6 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/sign_as_creditor.ts]: Done.

File: __test__/integration/loan_offer/runners/accept_as_debtor.ts (hunk 4 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/accept_as_debtor.ts]: Done.

File: __test__/integration/loan_offer/runners/accept_as_debtor.ts (hunk 5 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/accept_as_debtor.ts]: Done.

File: __test__/integration/loan_offer/runners/accept.ts (hunk 3 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/accept.ts]: Done.

File: __test__/integration/loan_offer/runners/accept.ts (hunk 2 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/accept.ts]: done.

File: __test__/integration/loan_offer/runners/accept.ts (hunk 1 above)
Comment (chrismin) [reply __test__/integration/loan_offer/runners/accept.ts]: done.

**Pull request merged**